File: backend/models/staff.py
"""
Staff Management Models
Handles staff profiles, service assignments, and availability
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import date, time
from decimal import Decimal
from uuid import UUID


# =====================================================
# WORKING HOURS MODELS
# =====================================================

class WorkingHoursDay(BaseModel):
    """Working hours for a single day"""
    enabled: bool = True
    start: str = Field(..., pattern=r"^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$")  # HH:MM format
    end: str = Field(..., pattern=r"^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$")
    breaks: List[Dict[str, str]] = []  # List of {start, end} break times
    
    @validator('end')
    def end_after_start(cls, v, values):
        if 'start' in values and v <= values['start']:
            raise ValueError('End time must be after start time')
        return v


class WorkingHours(BaseModel):
    """Complete weekly working hours"""
    monday: WorkingHoursDay
    tuesday: WorkingHoursDay
    wednesday: WorkingHoursDay
    thursday: WorkingHoursDay
    friday: WorkingHoursDay
    saturday: WorkingHoursDay
    sunday: WorkingHoursDay


# =====================================================
# STAFF MODELS
# =====================================================

class StaffBase(BaseModel):
    """Base staff model"""
    name: str = Field(..., min_length=2, max_length=100)
    bio: Optional[str] = None
    phone: Optional[str] = Field(None, pattern=r"^\+?[0-9]{10,15}$")
    email: Optional[str] = None
    working_hours: Optional[Dict[str, Any]] = None
    days_off: List[str] = []  # ISO date strings
    is_active: bool = True


class StaffCreate(StaffBase):
    """Create new staff member"""
    salon_id: UUID


class StaffUpdate(BaseModel):
    """Update staff member (all fields optional)"""
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    bio: Optional[str] = None
    phone: Optional[str] = Field(None, pattern=r"^\+?[0-9]{10,15}$")
    email: Optional[str] = None
    working_hours: Optional[Dict[str, Any]] = None
    days_off: Optional[List[str]] = None
    is_active: Optional[bool] = None


class StaffResponse(StaffBase):
    """Staff response with computed fields"""
    id: UUID
    salon_id: UUID
    image_url: Optional[str] = None
    average_rating: Decimal = Field(default=Decimal("0.00"))
    total_reviews: int = 0
    total_bookings: int = 0
    created_at: str
    updated_at: str
    
    class Config:
        from_attributes = True


class StaffWithServices(StaffResponse):
    """Staff with their assigned services"""
    services: List[Dict[str, Any]] = []


# =====================================================
# STAFF SERVICE ASSIGNMENT MODELS
# =====================================================

class StaffServiceAssignment(BaseModel):
    """Assign a service to a staff member"""
    staff_id: UUID
    service_id: UUID
    custom_price: Optional[Decimal] = Field(None, ge=0)
    custom_duration: Optional[int] = Field(None, gt=0)  # minutes


class StaffServiceResponse(BaseModel):
    """Staff service assignment response"""
    id: UUID
    staff_id: UUID
    service_id: UUID
    custom_price: Optional[Decimal] = None
    custom_duration: Optional[int] = None
    created_at: str
    
    class Config:
        from_attributes = True


class BulkStaffServiceAssignment(BaseModel):
    """Assign multiple services to a staff member at once"""
    staff_id: UUID
    service_ids: List[UUID] = Field(..., min_items=1)


# =====================================================
# STAFF AVAILABILITY MODELS
# =====================================================

class StaffAvailabilityCheck(BaseModel):
    """Check if staff is available for a specific time"""
    staff_id: UUID
    service_id: UUID
    booking_date: date
    time_slot: time
    duration: int = 30  # minutes


class AvailableStaffMember(BaseModel):
    """Staff member available for a specific service/time"""
    staff_id: UUID
    staff_name: str
    staff_image_url: Optional[str] = None
    staff_bio: Optional[str] = None
    average_rating: Decimal
    total_reviews: int
    custom_price: Optional[Decimal] = None
    custom_duration: Optional[int] = None


class AvailableStaffResponse(BaseModel):
    """Response with list of available staff"""
    salon_id: UUID
    service_id: UUID
    booking_date: date
    time_slot: time
    available_staff: List[AvailableStaffMember]
    any_available: bool  # True if at least one staff is available


# =====================================================
# STAFF PERFORMANCE MODELS
# =====================================================

class StaffPerformance(BaseModel):
    """Staff performance metrics"""
    staff_id: UUID
    staff_name: str
    total_bookings: int
    total_reviews: int
    average_rating: Decimal
    services_count: int
    revenue_generated: Optional[Decimal] = None
    
    class Config:
        from_attributes = True


class StaffStats(BaseModel):
    """Detailed staff statistics"""
    staff_id: UUID
    total_bookings: int
    completed_bookings: int
    cancelled_bookings: int
    total_revenue: Decimal
    average_rating: Decimal
    total_reviews: int
    most_booked_service: Optional[str] = None
    busiest_day: Optional[str] = None
    busiest_time_slot: Optional[str] = None


# =====================================================
# STAFF SCHEDULE MODELS
# =====================================================

class StaffScheduleDay(BaseModel):
    """Staff schedule for a single day"""
    date: date
    day_of_week: str
    is_working: bool
    working_hours: Optional[WorkingHoursDay] = None
    bookings: List[Dict[str, Any]] = []
    available_slots: List[str] = []  # List of available time slots


class StaffScheduleWeek(BaseModel):
    """Staff schedule for a week"""
    staff_id: UUID
    staff_name: str
    week_start: date
    week_end: date
    days: List[StaffScheduleDay]


# =====================================================
# STAFF IMAGE UPLOAD
# =====================================================

class StaffImageUpload(BaseModel):
    """Staff image upload response"""
    staff_id: UUID
    image_url: str
    uploaded_at: str


# =====================================================
# VALIDATION HELPERS
# =====================================================

def validate_working_hours(working_hours: Dict[str, Any]) -> bool:
    """Validate working hours structure"""
    required_days = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday']
    
    for day in required_days:
        if day not in working_hours:
            return False
        
        day_data = working_hours[day]
        if not isinstance(day_data, dict):
            return False
        
        if 'enabled' not in day_data or 'start' not in day_data or 'end' not in day_data:
            return False
    
    return True


def validate_time_format(time_str: str) -> bool:
    """Validate HH:MM time format"""
    try:
        hours, minutes = time_str.split(':')
        return 0 <= int(hours) <= 23 and 0 <= int(minutes) <= 59
    except:
        return False

File: mobile/src/types/staff.ts
/**
 * Staff Management Types
 * Comprehensive type definitions for staff selection system
 */

// =====================================================
// WORKING HOURS TYPES
// =====================================================

export interface WorkingHoursBreak {
  start: string; // HH:MM format
  end: string;   // HH:MM format
}

export interface WorkingHoursDay {
  enabled: boolean;
  start: string;  // HH:MM format
  end: string;    // HH:MM format
  breaks: WorkingHoursBreak[];
}

export interface WorkingHours {
  monday: WorkingHoursDay;
  tuesday: WorkingHoursDay;
  wednesday: WorkingHoursDay;
  thursday: WorkingHoursDay;
  friday: WorkingHoursDay;
  saturday: WorkingHoursDay;
  sunday: WorkingHoursDay;
}

// Default working hours template
export const DEFAULT_WORKING_HOURS: WorkingHours = {
  monday: { enabled: true, start: '09:00', end: '18:00', breaks: [] },
  tuesday: { enabled: true, start: '09:00', end: '18:00', breaks: [] },
  wednesday: { enabled: true, start: '09:00', end: '18:00', breaks: [] },
  thursday: { enabled: true, start: '09:00', end: '18:00', breaks: [] },
  friday: { enabled: true, start: '09:00', end: '18:00', breaks: [] },
  saturday: { enabled: true, start: '09:00', end: '18:00', breaks: [] },
  sunday: { enabled: false, start: '09:00', end: '18:00', breaks: [] },
};

// =====================================================
// STAFF TYPES
// =====================================================

export interface Staff {
  id: string;
  salon_id: string;
  name: string;
  bio?: string;
  image_url?: string;
  phone?: string;
  email?: string;
  working_hours: WorkingHours;
  days_off: string[]; // ISO date strings
  is_active: boolean;
  average_rating: number;
  total_reviews: number;
  total_bookings: number;
  created_at: string;
  updated_at: string;
}

export interface StaffWithServices extends Staff {
  services: StaffService[];
}

export interface StaffService {
  id: string;
  name: string;
  price: number;
  duration: number;
  description?: string;
  custom_price?: number;
  custom_duration?: number;
}

// =====================================================
// STAFF CREATION/UPDATE TYPES
// =====================================================

export interface StaffCreateInput {
  salon_id: string;
  name: string;
  bio?: string;
  phone?: string;
  email?: string;
  working_hours?: WorkingHours;
  days_off?: string[];
  is_active?: boolean;
}

export interface StaffUpdateInput {
  name?: string;
  bio?: string;
  phone?: string;
  email?: string;
  working_hours?: WorkingHours;
  days_off?: string[];
  is_active?: boolean;
}

// =====================================================
// STAFF SERVICE ASSIGNMENT TYPES
// =====================================================

export interface StaffServiceAssignment {
  staff_id: string;
  service_id: string;
  custom_price?: number;
  custom_duration?: number;
}

export interface BulkStaffServiceAssignment {
  staff_id: string;
  service_ids: string[];
}

// =====================================================
// STAFF AVAILABILITY TYPES
// =====================================================

export interface AvailableStaffMember {
  staff_id: string;
  staff_name: string;
  staff_image_url?: string;
  staff_bio?: string;
  average_rating: number;
  total_reviews: number;
  custom_price?: number;
  custom_duration?: number;
}

export interface AvailableStaffResponse {
  salon_id: string;
  service_id: string;
  booking_date: string;
  time_slot: string;
  available_staff: AvailableStaffMember[];
  any_available: boolean;
}

export interface StaffAvailabilityCheck {
  staff_id: string;
  service_id: string;
  booking_date: string;
  time_slot: string;
  duration?: number;
}

// =====================================================
// STAFF PERFORMANCE TYPES
// =====================================================

export interface StaffPerformance {
  staff_id: string;
  staff_name: string;
  total_bookings: number;
  total_reviews: number;
  average_rating: number;
  services_count: number;
  revenue_generated?: number;
}

export interface StaffStats {
  staff_id: string;
  total_bookings: number;
  completed_bookings: number;
  cancelled_bookings: number;
  total_revenue: number;
  average_rating: number;
  total_reviews: number;
  most_booked_service?: string;
  busiest_day?: string;
  busiest_time_slot?: string;
}

// =====================================================
// STAFF SELECTION UI TYPES
// =====================================================

export interface StaffSelectionOption {
  type: 'any' | 'specific';
  staff_id?: string;
  staff_name?: string;
  staff_image_url?: string;
  average_rating?: number;
  total_reviews?: number;
  custom_price?: number;
  custom_duration?: number;
}

// Special "Any Available" option
export const ANY_STAFF_OPTION: StaffSelectionOption = {
  type: 'any',
  staff_name: 'Any Available Staff',
};

// =====================================================
// HELPER FUNCTIONS
// =====================================================

/**
 * Check if a staff member is working on a specific day
 */
export function isStaffWorkingOnDay(
  staff: Staff,
  date: Date
): boolean {
  const dayName = date.toLocaleDateString('en-US', { weekday: 'long' }).toLowerCase() as keyof WorkingHours;
  const daySchedule = staff.working_hours[dayName];
  
  // Check if day is enabled
  if (!daySchedule.enabled) {
    return false;
  }
  
  // Check if date is in days_off
  const dateString = date.toISOString().split('T')[0];
  if (staff.days_off.includes(dateString)) {
    return false;
  }
  
  return true;
}

/**
 * Get working hours for a specific day
 */
export function getWorkingHoursForDay(
  staff: Staff,
  date: Date
): WorkingHoursDay | null {
  if (!isStaffWorkingOnDay(staff, date)) {
    return null;
  }
  
  const dayName = date.toLocaleDateString('en-US', { weekday: 'long' }).toLowerCase() as keyof WorkingHours;
  return staff.working_hours[dayName];
}

/**
 * Format staff rating for display
 */
export function formatStaffRating(rating: number): string {
  return rating.toFixed(1);
}

/**
 * Get staff display name with rating
 */
export function getStaffDisplayName(staff: Staff | AvailableStaffMember): string {
  const name = 'name' in staff ? staff.name : staff.staff_name;
  const rating = 'average_rating' in staff ? staff.average_rating : 0;
  const reviews = 'total_reviews' in staff ? staff.total_reviews : 0;
  
  if (reviews > 0) {
    return `${name} (${formatStaffRating(rating)} ⭐)`;
  }
  
  return name;
}

/**
 * Get effective price for a service with staff
 */
export function getEffectivePrice(
  basePrice: number,
  customPrice?: number
): number {
  return customPrice ?? basePrice;
}

/**
 * Get effective duration for a service with staff
 */
export function getEffectiveDuration(
  baseDuration: number,
  customDuration?: number
): number {
  return customDuration ?? baseDuration;
}

/**
 * Calculate price difference for staff premium/discount
 */
export function getPriceDifference(
  basePrice: number,
  customPrice?: number
): { amount: number; percentage: number; type: 'premium' | 'discount' | 'same' } {
  if (!customPrice || customPrice === basePrice) {
    return { amount: 0, percentage: 0, type: 'same' };
  }
  
  const difference = customPrice - basePrice;
  const percentage = (difference / basePrice) * 100;
  
  return {
    amount: Math.abs(difference),
    percentage: Math.abs(percentage),
    type: difference > 0 ? 'premium' : 'discount',
  };
}

/**
 * Validate phone number format
 */
export function isValidPhone(phone: string): boolean {
  const phoneRegex = /^\+?[0-9]{10,15}$/;
  return phoneRegex.test(phone);
}

/**
 * Validate time format (HH:MM)
 */
export function isValidTimeFormat(time: string): boolean {
  const timeRegex = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/;
  return timeRegex.test(time);
}

/**
 * Compare two time strings (HH:MM format)
 * Returns: -1 if time1 < time2, 0 if equal, 1 if time1 > time2
 */
export function compareTimeStrings(time1: string, time2: string): number {
  const [h1, m1] = time1.split(':').map(Number);
  const [h2, m2] = time2.split(':').map(Number);
  
  if (h1 < h2) return -1;
  if (h1 > h2) return 1;
  if (m1 < m2) return -1;
  if (m1 > m2) return 1;
  return 0;
}

/**
 * Get day name from date
 */
export function getDayName(date: Date): keyof WorkingHours {
  return date.toLocaleDateString('en-US', { weekday: 'long' }).toLowerCase() as keyof WorkingHours;
}

/**
 * Check if time is within working hours
 */
export function isTimeWithinWorkingHours(
  time: string,
  workingHours: WorkingHoursDay
): boolean {
  if (!workingHours.enabled) {
    return false;
  }
  
  return (
    compareTimeStrings(time, workingHours.start) >= 0 &&
    compareTimeStrings(time, workingHours.end) < 0
  );
}
